crates/ironrdp-pdu/src/lib.rs

```diff
@@ -148,11 +148,21 @@ pub fn encode<T: PduEncode>(pdu: &T, dst: &mut [u8]) -> PduResult<usize> {
     Ok(cursor.pos())
 }
 
+pub fn encode_box(pdu: Box<dyn PduEncode>, dst: &mut [u8]) -> PduResult<usize> {
+    let mut cursor = WriteCursor::new(dst);
+    encode_cursor_box(pdu, &mut cursor)?;
+    Ok(cursor.pos())
+}
```

Comment:
If I understand correctly, we don’t need all the `*_box` functions. The following code should compile fine:
```rust
fn chunkify_one(pdu: Box<dyn PduEncode>) -> PduResult<Vec<WriteBuf>> {
    let mut encoded_pdu = WriteBuf::new();
    encode_buf(&pdu, &mut encoded_pdu)?; // <-- `&buf`
    // If it doesn’t, maybe `as_ref` will return what we want
    encode_buf(pdu.as_ref(), &mut encoded_pdu)?; // <-- `buf.as_ref()`
    // …
}
```

Reply:
For `encode_buf(&pdu, &mut encoded_pdu)?;`, I get the error:

```
error[E0277]: the trait bound `std::boxed::Box<dyn ironrdp_pdu::PduEncode>: ironrdp_pdu::mcs::McsPdu<'_>` is not satisfied
   --> crates/ironrdp-svc/src/lib.rs:80:16
    |
80  |     encode_buf(&pdu, &mut encoded_pdu)?;
    |     ---------- ^^^^ the trait `ironrdp_pdu::mcs::McsPdu<'_>` is not implemented for `std::boxed::Box<dyn ironrdp_pdu::PduEncode>`
    |     |
    |     required by a bound introduced by this call
    |
    = help: the following other types implement trait `ironrdp_pdu::mcs::McsPdu<'de>`:
              ironrdp_pdu::mcs::AttachUserConfirm
              ironrdp_pdu::mcs::AttachUserRequest
              ironrdp_pdu::mcs::ChannelJoinConfirm
              ironrdp_pdu::mcs::ChannelJoinRequest
              ironrdp_pdu::mcs::DisconnectProviderUltimatum
              ironrdp_pdu::mcs::ErectDomainPdu
              ironrdp_pdu::mcs::McsMessage<'de>
              ironrdp_pdu::mcs::SendDataIndication<'de>
              ironrdp_pdu::mcs::SendDataRequest<'de>
    = note: required for `std::boxed::Box<dyn ironrdp_pdu::PduEncode>` to implement `ironrdp_pdu::x224::X224Pdu<'_>`
    = note: required for `std::boxed::Box<dyn ironrdp_pdu::PduEncode>` to implement `ironrdp_pdu::PduEncode`
note: required by a bound in `ironrdp_pdu::encode_buf`
   --> /Users/ibeckermayer/Devolutions/IronRDP/crates/ironrdp-pdu/src/lib.rs:168:22
    |
168 | pub fn encode_buf<T: PduEncode>(pdu: &T, buf: &mut WriteBuf) -> PduResult<usize> {
    |                      ^^^^^^^^^ required by this bound in `encode_buf`
```

It's not exactly clear to me why it's complaining about `McsPdu` not being implemented when `PduEncode` is what's required by `encode_buf`, though I think it may be related to the chain of

```rust
impl<'de, T> X224Pdu<'de> for T where T: McsPdu<'de>;
impl<'de, T> PduEncode for T where T: X224Pdu<'de>;
```

---

For `encode_buf(pdu.as_ref(), &mut encoded_pdu)?;`, I get

```
error[E0277]: the size for values of type `dyn ironrdp_pdu::PduEncode` cannot be known at compilation time
   --> crates/ironrdp-svc/src/lib.rs:80:16
    |
80  |     encode_buf(pdu.as_ref(), &mut encoded_pdu)?;
    |     ---------- ^^^^^^^^^^^^ doesn't have a size known at compile-time
    |     |
    |     required by a bound introduced by this call
    |
    = help: the trait `std::marker::Sized` is not implemented for `dyn ironrdp_pdu::PduEncode`
note: required by a bound in `ironrdp_pdu::encode_buf`
   --> /Users/ibeckermayer/Devolutions/IronRDP/crates/ironrdp-pdu/src/lib.rs:168:19
    |
168 | pub fn encode_buf<T: PduEncode>(pdu: &T, buf: &mut WriteBuf) -> PduResult<usize> {
    |                   ^ required by this bound in `encode_buf`
```

but I can't make `PduEncode: Sized` because that breaks its object safety.

Reply:
I think the `McsPdu` error message might just be a weird consequence of how the compiler checks whether `PduEncode` is implemented. E.g. it checks whether `PduEncode` is implemented, then if not it sees `impl<'de, T> PduEncode for T where T: X224Pdu<'de>;` so checks whether `X224Pdu` is implemented, then since that's not the case it knows `impl<'de, T> X224Pdu<'de> for T where T: McsPdu<'de>;` and so checks if `McsPdu`. And since all of those fail, it fails, but it only reports the last one it checked. If that's the case then that's confusing and should probably be fixed by the rust team.

Anyways, I managed to get rid of the `*_box()` functions in 868d7ec962a5de7962f7c08e04e310db04f4bf86

Reply:
Ah okay! Now that I see the error, I understand. Actually, there is a trait bound, `Sized`, that is implicit for all type parameters. I believe the `encode_*` function family should opt out of this [using `?Sized`](https://doc.rust-lang.org/std/marker/trait.Sized.html).
Your approach works by taking a `Box` , which by itself is a `Sized` type, pointing to a dynamically sized type, a trait object.

Reply:
I see, good find: e31467f45515f17120c1508e9f091a35fe28a534

Reply:
I see! Good find. 

e31467f45515f17120c1508e9f091a35fe28a534